binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java

```diff
@@ -42,18 +44,42 @@ private ServerSecurityPolicy(ImmutableMap<String, SecurityPolicy> perServicePoli
   }
 
   /**
-   * Return whether the given Android UID is authorized to access a particular service.
+   * Returns whether the given Android UID is authorized to access a particular service.
    *
-   * <b>IMPORTANT</b>: This method may block for extended periods of time.
+   * <p><b>IMPORTANT</b>: This method may block for extended periods of time.
    *
    * @param uid The Android UID to authenticate.
    * @param serviceName The name of the gRPC service being called.
+   * @deprecated Application code should not need to call this method.
    */
   @CheckReturnValue
+  @Deprecated
   public Status checkAuthorizationForService(int uid, String serviceName) {
     return perServicePolicies.getOrDefault(serviceName, defaultPolicy).checkAuthorization(uid);
   }
 
+  /**
+   * Returns whether the given Android UID is authorized to access a particular service.
+   *
+   * <p>This method never throws an exception. If the execution of the security policy check
+   * fails, a failed future with such exception is returned.
+   *
+   * @param uid The Android UID to authenticate.
+   * @param serviceName The name of the gRPC service being called.
+   * @return a future with the result of the authorization check. A failed future represents a
+   *     failure to perform the authorization check, not that the access is denied.
+   */
+  @CheckReturnValue
+  ListenableFuture<Status> checkAuthorizationForServiceAsync(int uid, String serviceName) {
```

Comment:
For the operation to be async, should we pass in an executor here? We do that in similar APIs. That would allow this adapter implementation to be async as well.

Not necessary for this PR (since this is all internal API), but something to consider.

Reply:
If we do that, that means the gRPC library is responsible for picking which executor to pass in. And it may not pick the appropriate one without knowing in advance what the implementation intends to do.

What I want here is to give implementers to use the executor of their choice. For instance, Google's `@BlockingExecutor`or `@BackgroundExecutor` depending on the type of blocking call they are issuing. 

From the library's perspective, all that matters is clients trying to be well-behaved and avoiding blocking the gRPC thread, but they are free to choose how to write their async code.

Reply:
gRPC already has an executor for blocking operations:
https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html#offloadExecutor(java.util.concurrent.Executor)

It is used for DNS and retrieving oauth tokens. That sort of work.

Reply:
But not all slow calls are the same. go/which-executor, for instance, distinguishes those that scale well (-> `@BlockingExecutor`; eg: network calls) from those that do not (-> `@BackgroundExecutor`; eg: CPU-intensive work, IPC, disk I/O etc.). We don't know which kind will be implemented.

There is a future CL in which I do plan to use an executor internally to handle the resulting `Status` object. For now, however, I think it's cleaner to leave it up to implementers and not suggest any particular executor. They are still welcome to use the same one they passed to the channel builder, but the framework is agnostic to that.